pkg/daemon/multus/validation.go

```diff
@@ -0,0 +1,314 @@
+	if err != nil {
+		return []string{"inability to schedule DaemonSets is likely an issue with the Kubernetes cluster itself"},
+			fmt.Errorf("expected number of client not yet ready: %w", err)
+	}
+	vsm.vt.Logger.Infof("expecting %d clients", expectedNumClientPods)
+	vsm.SetNextState(&verifyAllClientsRunningState{
+		expectedNumClients: expectedNumClientPods,
+	})
+	return []string{}, nil
+}
+
+// 4. All client pods should be in running state, even if they aren't "Ready"
+type verifyAllClientsRunningState struct {
+	expectedNumClients int
+}
+
+func (s *verifyAllClientsRunningState) Run(ctx context.Context, vsm *validationStateMachine) (suggestions []string, err error) {
+	running, err := vsm.vt.allClientsAreRunning(ctx, s.expectedNumClients)
+	suggestions = append([]string{
+		"clients not being able to run can mean multus is unable to provide them with addresses"},
+		unableToProvideAddressSuggestions...)
+	if err != nil {
+		return suggestions, err
+	} else if !running {
+		return suggestions, nil // not running, but no specific error
+	}
+	vsm.vt.Logger.Infof("all %d clients are running - but may not be ready", s.expectedNumClients)
+	vsm.SetNextState(&verifyAllClientsReadyState{
+		expectedNumClients: s.expectedNumClients,
+	})
+	return []string{}, nil
+}
+
+// 5. All client pods should be in "Ready" state
+type verifyAllClientsReadyState struct {
+	expectedNumClients int
+
+	// keep some info to heuristically determine if the network might be flaky/overloaded
+	prevNumReady                    int
+	timeClientsStartedBecomingReady time.Time
+	suggestFlaky                    bool
+}
+
+func (s *verifyAllClientsReadyState) Run(ctx context.Context, vsm *validationStateMachine) (suggestions []string, err error) {
+	numReady, err := vsm.vt.numClientsReady(ctx, s.expectedNumClients)
+	collocationSuggestion := "if clients on the same node as the web server become ready but not others, " +
+		"there may be a network firewall or security policy blocking inter-node traffic on multus networks"
+	defaultSuggestions := append([]string{collocationSuggestion, flakyNetworkSuggestion}, unableToProvideAddressSuggestions...)
+	if err != nil {
+		return defaultSuggestions, err
+	}
+
+	s.checkIfFlaky(vsm, numReady)
+
+	if numReady != s.expectedNumClients {
+		return defaultSuggestions, fmt.Errorf("number of ready clients [%d] is not the number expected [%d]", numReady, s.expectedNumClients)
+	}
+
+	vsm.vt.Logger.Infof("all %d clients are ready", s.expectedNumClients)
+	suggestionsOnSuccess := []string{}
+	if s.suggestFlaky {
+		suggestionsOnSuccess = append(suggestionsOnSuccess,
+			fmt.Sprintf("not all clients became ready within %s; %s", flakyThreshold.String(), flakyNetworkSuggestion))
+	}
+	vsm.Exit() // DONE!
+	return suggestionsOnSuccess, nil
+}
+
+// clients should all become ready within a pretty short amount of time since they all should start
+// pretty simultaneously
+var flakyThreshold = 20 * time.Second
```

Comment:
If any image is being pulled, we may need longer than this even if the image is small.

Reply:
That's true. A good future improvement would be to allow this to be tuned also, or for the test to ensure the binary is pulled on all nodes first. Because of the multiple ways of moving forward, I figured it would be good to defer for the future.